### PR TITLE
WIP: Added initial support for qnx

### PR DESCRIPTION
**Pull request by deepakchethan**
For issue #12664 

This is still not complete
I did the base part and have some questions

1. What should I do with the fixup records?
2. Are the addresses used correctly?

### CODE REVIEW

File: libr/bin/format/qnx/qnx.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include <r_lib.h>
+#include <r_types.h>
+#include <r_bin.h>
+
+enum {
+    LMF_HEADER_REC = 0,
+    LMF_COMMENT_REC,
+    LMF_LOAD_REC,
+    LMF_FIXUP_REC,
+    LMF_8087_FIXUP_REC,
```
Comment (radare) [inline libr/bin/format/qnx/qnx.h]: use tabs

File: libr/bin/p/bin_qnx.c [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+			}
+			ptr->name = strdup ("LMF_LOAD");
+			ptr->paddr = offset;
+			ptr->vaddr = ldata->offset;
+			ptr->vsize = lrec->data_nbytes - sizeof (lmf_data);
+			ptr->size = ptr->vsize;
+			ptr->add = true;
+		 	r_list_append (ret, ptr);
+		}
+        else if (lrec->rec_type == LMF_RW_END_REC) {
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: bad indentation , the } must be in the same line as the else

File: libr/bin/p/bin_qnx.c [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+		free (qo);
+		return false;
+	}
+	// Read the first record
+	r_buf_fread_at (bf->buf, 0, (ut8 *) lrec, "iiii", 1);
+	// Load the header
+	lmf_header_load (&qo->lmfh, bf->buf, qo->kv);
+	offset += lrec->data_nbytes;
+	
+	for( ;; ) {
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: bad indentation

File: libr/bin/p/bin_qnx.c [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+		(void) r_buf_fread_at (bf->buf, offset, (ut8 *) lrec, "iiii", 1);
+		offset += sizeof( lmf_record );
+
+        if( lrec->rec_type == LMF_IMAGE_END_REC ) {
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: bad indentation and spacings

File: libr/bin/p/bin_qnx.c [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            break;
+        }
+		else if (lrec->rec_type == LMF_RESOURCE_REC) {
+			RBinSection *ptr = NULL;
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: R_NEW0() in here, instead of assign to null

File: libr/bin/p/bin_qnx.c [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+			r_buf_fread_at (bf->buf, offset, (ut8 *) lres, "iccc", 1);
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: maybe better to check the return value and see if we have read all the bytes

File: libr/bin/p/bin_qnx.c [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+			}
+			ptr->name = strdup ("LMF_RESOURCE");
+			ptr->paddr = offset;
+			ptr->vsize = lrec->data_nbytes - sizeof (lmf_resource);
+			ptr->size = ptr->vsize;
+			ptr->add = true;
+		 	r_list_append (ret, ptr);
+		}
+		else if (lrec->rec_type == LMF_LOAD_REC) {
+			RBinSection *ptr = NULL;
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: same here

File: libr/bin/p/bin_qnx.c [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+			ptr->vsize = lrec->data_nbytes - sizeof (lmf_data);
+			ptr->size = ptr->vsize;
+			ptr->add = true;
+		 	r_list_append (ret, ptr);
+		}
+        else if (lrec->rec_type == LMF_RW_END_REC) {
+			r_buf_fread_at (bf->buf, offset, (ut8 *) &qo->rwend, "ii", 1);
+		}
+        offset += lrec->data_nbytes;
+    }
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: indent

File: libr/bin/p/bin_qnx.c [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+	sdb_ns_set (sdb, "info", qo->kv);
+	qo->sections = ret;
+	*bin_obj = qo;
+	return true;
+}
+
+static bool load(RBinFile *bf) {
+	const ut8 *byte = bf? r_buf_buffer (bf->buf): NULL;
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: dont do that. r_buf_buffer is not recommended, because it doesnt control boundaries

File: libr/bin/p/bin_qnx.c [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+	ut64 size = bf? r_buf_size (bf->buf): 0;
+	if (!bf || !bf->o) {
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: put that check on top of the function but use r_return_val_if_fail (bf && bf->o, false);

File: libr/bin/p/bin_qnx.c [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+		return false;
+	}
+	return load_bytes (bf, &bf->o->bin_obj, byte, size, bf->o->loadaddr, bf->sdb);
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: is better to just use RBuffer around, and remove the use of "byte + size"

File: libr/bin/p/bin_qnx.c [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+/*
+ * Provides the info about the binary file
+ * @param RBinFile to extract the data from
+ * @return RBinInfo file with the info
+ */
+static RBinInfo *info(RBinFile *bf) {
+	RBinInfo *ret;
+	if (!bf || !bf->o || !bf->o->bin_obj) {
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: use r_return_val_if_fail

File: libr/bin/p/bin_qnx.c [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+		return NULL;
+	}
+	ret = R_NEW0 (RBinInfo);
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: define and assign in the same line

File: libr/bin/p/bin_qnx.c [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+	ret->lang = strdup ("C/C++");
+	ret->signature = true;
+	return ret;
+}
+
+static void header(RBinFile *bf) {
+	QnxObj *bin = bf->o->bin_obj;
+	struct r_bin_t *rbin = bf->rbin;
+
+	rbin->cb_printf ("QNX file header:\n");
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: too much spaces

File: libr/bin/p/bin_qnx.c [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+	ret->os = strdup ("any");
+	ret->subsystem = strdup ("any");
+	ret->lang = strdup ("C/C++");
+	ret->signature = true;
+	return ret;
+}
+
+static void header(RBinFile *bf) {
+	QnxObj *bin = bf->o->bin_obj;
+	struct r_bin_t *rbin = bf->rbin;
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: use RBin instead of struct r_bin_t

File: libr/bin/p/bin_qnx.c [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+// Returns the sections 
+static RList* sections(RBinFile *bf) {
+	QnxObj *qo = bf->o->bin_obj;
+	RList *ret = NULL;
+	if (!(ret = r_list_new ())) {
+		return NULL;
+	}
+	ret = r_list_clone (qo->sections);
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: just `return r_list_clone (qo->sections);`

... are you sure you want to clone the list?

File: libr/bin/p/bin_qnx.c [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+	QnxObj *qo = bf->o->bin_obj;
+	if (!(ret = r_list_new ())) {
+		return NULL;
+	}
+	ret->free = free;
+	if (!(ptr = R_NEW0 (RBinAddr))) {
+		return ret;
+	}
+	ptr->paddr = qo->lmfh.code_offset;
+	ptr->vaddr = qo->lmfh.code_offset;
```
Comment (radare) [inline libr/bin/p/bin_qnx.c]: baddr = 0?

**Comment by radare** [comment]
What do you mean by fixuo the records?

File: libr/bin/format/qnx/qnx.h (hunk 1 above)
Comment (deepakchethan) [reply libr/bin/format/qnx/qnx.h]: Done

File: libr/bin/p/bin_qnx.c (hunk 2 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

File: libr/bin/p/bin_qnx.c (hunk 4 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

File: libr/bin/p/bin_qnx.c (hunk 3 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

File: libr/bin/p/bin_qnx.c (hunk 5 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Done

File: libr/bin/p/bin_qnx.c (hunk 8 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

File: libr/bin/p/bin_qnx.c (hunk 10 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Done

File: libr/bin/p/bin_qnx.c (hunk 11 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: I might be wrong but I think`load_bytes` method must have all these args.  If I change here I might have to change it in `r_bin_object_new` and in turn have to change it in all plugins.

I will wait for your input and then proceed with this.

File: libr/bin/p/bin_qnx.c (hunk 12 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Done

File: libr/bin/p/bin_qnx.c (hunk 13 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

File: libr/bin/p/bin_qnx.c (hunk 14 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Done

File: libr/bin/p/bin_qnx.c (hunk 17 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Added the `baddr` to `ptr->vaddr` based on one of the plugin.
This was my question actually. I am confused with `paddr` and `vaddr`. None of the places on the internet clearly defines how to calculate them.

**Comment by deepakchethan** [comment]
> What do you mean by fixuo the records?
> […](#)
Even these are a type of records in the binary. Even these use the `qnx_data` struct. My doubt is, should I just add into the sections or relocs.
In Wikipedia fixup is defined as part of relocs. But it doesn't have enough stuff to use RBinReloc

File: libr/bin/p/bin_qnx.c (hunk 15 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

File: libr/bin/p/bin_qnx.c (hunk 9 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Okay, I will use `r_buf_get_at` instead.

Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: Okay, I will use `r_buf_get_at` instead.

File: libr/bin/p/bin_qnx.c (hunk 16 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: If I don't clone the list, the destroy routine will cause a segmentation error.

File: libr/bin/p/bin_qnx.c (hunk 6 above)
Comment (deepakchethan) [reply libr/bin/p/bin_qnx.c]: done

**Comment by Maijin** [comment]
Feel free to update the README fileformat ;) https://github.com/radare/radare2#file-formats

**Comment by radare** [comment]
Can we merge this, add some tests and continue in a separate PR?

**Comment by deepakchethan** [comment]
> Can we merge this, add some tests and continue in a separate PR?
Yeah, that sounds good :)

**Comment by radare** [comment]
awesome! thanks!!

**Pull request merged**